### ISSUE TITLE
Suppression for long identifiers for `4.4 Column Limit: 100` rule of google java style guide

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15233]
I have read check documentation: https://checkstyle.org/checks/sizes/linelength.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Detected at: https://github.com/checkstyle/checkstyle/pull/14901#discussion_r1641831677

The [4.4 Column Limit: 100](https://google.github.io/styleguide/javaguide.html#s4.4-column-limit) rule of google java style guide has some exceptions cases in which this rule does not apply, one of them is: 

> Very long identifiers, on the rare occasions they are called for, are allowed to exceed the column limit. In that case, the valid wrapping for the surrounding code is as produced by [google-java-format](https://github.com/google/google-java-format).

style guide says that long identifiers which crosses the column limit of 100 are allowed, so we should not give violation due to such identifiers. 

---

**Current behavior of Checkstyle:**

```java
package com.examples.columnlimit;

/** testing. */
public class Checkstyle {
  int aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa = 99;

  public void testtttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt() {
    System.out.println("heloooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo");
  }
}
```

**Output:**

```
$ java -jar ..\checkstyle-10.17.0-all.jar -c ..\google_checks.xml .\Checkstyle.java
Starting audit...
[WARN] \Checkstyle.java:5: Line is longer than 100 characters (found 109). [LineLength]
[WARN] \Checkstyle.java:7: Line is longer than 100 characters (found 106). [LineLength]
[WARN] \Checkstyle.java:8: Line is longer than 100 characters (found 110). [LineLength]
Audit done.
```

**google-java-format behavior:**

```java
package com.examples.columnlimit;

/** testing. */
public class Checkstyle {
  int
      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa =
          99;

  public void
      testtttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt() {
    System.out.println(
        "heloooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo");
  }
}
```

---

The formatter formats and tries to wrap the long identifier to make it under column limit but this still crosses the limit of 100. And in case of CS, we give violation for such type of code.

---

**Describe what you expect in detail.**

As @rdiachenko suggested here: https://github.com/checkstyle/checkstyle/pull/14901#discussion_r1660884679, we should revisit the working of [LineLength](https://checkstyle.org/checks/sizes/linelength.html) module and add a new property to ignore long identifiers, this will also require to update the [google_checks.xml](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml) accordingly. 

Or 

We can discuss another way to solve the problem here in this issue.

**Comment by romani** [comment]
@Zopsss , please make PR with Input that shows shows long identifiers wrapped by formatter as he can AND how checkstyle violate them.

Long variable, method, type name, ... Give a bit more examples. Keep it formatted by formatter.

**Comment by romani** [comment]
> we should revisit the working of [LineLength](https://checkstyle.org/checks/sizes/linelength.html) module and add a new property to ignore long identifiers,

it is not that easy, LineLength become not Java Check, so it does not know anything about methods and variables.

points of attention `....aaaaaaaaaaa =` and `...tttttttttt() {`
so tool does not break other rules for "="  and `{` to do line wrapping.

**Comment by Zopsss** [comment]
> @Zopsss , please make PR with Input that shows shows long identifiers wrapped by formatter as he can AND how checkstyle violate them.
> 
> Long variable, method, type name, ... Give a bit more examples. Keep it formatted by formatter. attention to example with a lot of nested code and lond identifier (20 characters) in most deep point of it. So it is not very long but still line with 100 can not fit it.

Umm creating a separate PR won't help much, I think it will be better to show the code here with the output, I guess this will be a more efficient way.

**Original code:**

```java
/** some javadoc. */
public class TestingColumnLimit {
  public static int GLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOBBBBBBBBAAAAAAAAALLLLLLLLLLL_______VARRRRRRRIIIIIIIABLEEEEEEEEE = 99;

  final int abccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc = 99;

  void fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo() {

    boolean xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx = true;

    if (xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx) {
      // some code
    } else {
      // some code
    }

    switch (GLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOBBBBBBBBAAAAAAAAALLLLLLLLLLL_______VARRRRRRRIIIIIIIABLEEEEEEEEE) {
        case abccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc:
            fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo();
            break;

        default:
            break;
    }

    System.out.println("Helllllllllllllllllllllllllllllllllllllllllllllllllllllllllloooooooooooooooooooooooooooooooooo");
  }

  class Innnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnneeeeeeeeeeeeeeeeeeeeeeeerrrrrrrrrrrrrrrrrrrr {}
}
```

**Formatted code:**

```java
/** some javadoc. */
public class TestingColumnLimit {
  public static int
      GLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOBBBBBBBBAAAAAAAAALLLLLLLLLLL_______VARRRRRRRIIIIIIIABLEEEEEEEEE =
          99;

  final int
      abccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc =
          99;

  void
      fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo() {

    boolean
        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx =
            true;

    if (xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx) {
      // some code
    } else {
      // some code
    }

    switch (GLLLLLLLLLLLLLLLLLLOOOOOOOOOOOOOOOBBBBBBBBAAAAAAAAALLLLLLLLLLL_______VARRRRRRRIIIIIIIABLEEEEEEEEE) {
      case abccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc:
        fooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo();
        break;

      default:
        break;
    }

    System.out.println(
        "Helllllllllllllllllllllllllllllllllllllllllllllllllllllllllloooooooooooooooooooooooooooooooooo");
  }

  class Innnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnneeeeeeeeeeeeeeeeeeeeeeeerrrrrrrrrrrrrrrrrrrr {}
}
```

**Output:**

```
$ java -jar .\checkstyle-10.17.0-all.jar -c .\google_checks.xml .\TestingColumnLimit.java
Starting audit...
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:4: Line is longer than 100 characters (found 105). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:8: Line is longer than 100 characters (found 111). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:12: Line is longer than 100 characters (found 106). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:15: Line is longer than 100 characters (found 106). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:18: Line is longer than 100 characters (found 107). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:24: Line is longer than 100 characters (found 112). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:25: Line is longer than 100 characters (found 115). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:26: Line is longer than 100 characters (found 107). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:34: Line is longer than 100 characters (found 106). [LineLength]
[WARN] \TestingGoogleStyle\.\TestingColumnLimit.java:37: Line is longer than 100 characters (found 122). [LineLength]
Audit done.
```

Formatter tries to shrink down the line size but it still crosses the limit of 100.

**Comment by Zopsss** [comment]
> > we should revisit the working of [LineLength](https://checkstyle.org/checks/sizes/linelength.html) module and add a new property to ignore long identifiers,
> 
> it is not that easy, LineLength become not Java Check, so it does not know anything about methods and variables.
> 
> points of attention `....aaaaaaaaaaa =` and `...tttttttttt() {` so tool does not break other rules for "=" and `{` to do line wrapping.
> 
> The only solution I see is to allow users to use suppressions to mark suppression in comment and explain reason https://checkstyle.org/google_style.html#Google_Suppressions, please play with this and lets choose what we would recomment for users to use if they run into this.

I think the [SuppressWarningsFilter](https://checkstyle.org/filters/suppresswarningsfilter.html#SuppressWarningsFilter) would be a good choice to recommend it to users, users' will just need to use `SuppressWarnings` annotation where they need suppression, something like:

```
  @SuppressWarnings({"LineLength"})
  final int abccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc = 99;
```

**Comment by romani** [comment]
@rdiachenko , do you agree with suppression?

**Comment by Zopsss** [comment]
@rdiachenko

**Comment by romani** [comment]
@Zopsss , while we are waiting for @rdiachenko , please send PR with update of Inputs annotations to show how it will looks like for user.
Please also update coverage table to mention that user should use extra code update to avoid violation on such edge cases.

**Comment by Zopsss** [comment]
done https://github.com/checkstyle/checkstyle/pull/15460 @romani

**Comment by rdiachenko** [comment]
> LineLength become not Java Check, so it does not know anything about methods and variables

I missed this point. Makes sense. I'm good with using suppression.
